### ISSUE TITLE
disassociated ownership

### ISSUE THREAD

**Issue by devsnek** [rricard/proposal-refcollection#7]
Some "basic" usage of ref collections and records:

```js
// We have some records and refs (which must always be tracked together)
let refA = ...;
let recA = ...;

let refB = ...;
let recB = ...;

// We combine them
let refC = ...;
let recC = #{ recA, recB, ... };
// OR ...
let refD = ...;
let recD = #{ ...recA, ...recB, ... };
```

There are a number of problems present here.

1. refX and recX must *always* be tracked together.
2. The value representing a ref is not inherently a ref, it's a symbol. There's no way for someone to know which symbols need dereferencing.
3. When you combine records you end up tracking ~~N+1 ref collections (where N is the number of records you combined)~~ an exponential amount of ref collections (because you inherit all the collections those refs were tracked with) without any way of knowing which symbols go to which references, and now you have to drag around all these ref collections for 1 record.
4. The semantics of combining ref collections to fix (3) are most likely undecidable because the old ref collection still has to exist and will likely be used by other parts of the code, resulting in symbol reuse between collections.

**Comment by jridgewell** [comment]
> The value representing a ref is not inherently a ref, it's a symbol. There's no way for someone to know which symbols need dereferencing.

When I originally spoke with them, it was impossible to insert a Symbol directly into a Record. So if it a record contains symbol, it's a ref symbol.

**Comment by devsnek** [comment]
@jridgewell i'm not aware of that limitation, only thing the proposal mentions is that record keys can't be symbols.

**Comment by jridgewell** [comment]
This was ~2-3 months ago, so it could have changed recently. In the discussion, I had noted that I used numbers as my sigils (and prevented you from directly using numbers), and then we symbols were suggested.

**Comment by ljharb** [comment]
If only some kinds of symbols are permitted in a Record, i have to have a reflection mechanism to determine if a symbol is a “ref symbol”. What method would that be?

**Comment by rickbutton** [comment]
To clarify: it is possible to use any symbol as a property value in a record or tuple, and not possible to use a symbol as a property key in a record.

**Comment by rickbutton** [comment]
@devsnek 

On 1, that is correct, you must keep the `RefCollection` alive to keep the object references alive. Alternatively, instead of creating a `RefCollection` per record, you can create a single `RefCollection` and store all references in that `RefCollection`, to the same effect, if you don't want to manage the lifetime of the `RefCollection`.

On 2, you can use a symbol's presence in the `RefCollection` to indicate that it needs to be dereferenced.

On 3, You can use the solution I mentioned for 1 here as well, maintain a single `RefCollection` instead.

**Comment by devsnek** [comment]
@rickbutton

> Alternatively, instead of creating a RefCollection per record, 

the records might come from anywhere. what will the pattern be? Do libraries expect you to pass a ref collection? Do libraries give you a ref collection? This seems rather chaotic.

> you can use a symbol's presence in the RefCollection to indicate that it needs to be dereferenced.

So you have to just exhaustively check every time you encounter a symbol?

> that is correct, you must keep the RefCollection alive to keep the object references alive

I don't mean lifetime, I mean you must literally carry one around wherever you carry the other around in order to be able to access all the values in the record.

**Comment by rricard** [comment]
A big solution space **not explored yet** in that proposal is to introduce a realm-wide RefCollection accessible from some namespace. This would be the most common usage case and you wouldn't need to move around any RefCollection.

As for having some way of having self-reflection in symbols to determine their RefCollection of origin, I'm not opposed, I'd just like to know what would be possible to do there...

**Comment by devsnek** [comment]
if there's one global collection, and by default everyone uses it to put objects in records, and symbols representing objects in records have identical identity semantics to having the objects in the records, why not just put the objects in the records. the argument seems to be "to make you have to dereference so you remember its an object" but i don't think we should use pain as a reminder.

**Comment by rricard** [comment]
> the argument seems to be "to make you have to dereference so you remember its an object" but that seems to be pretty obvious

I mean it's the main argument to me, there are others that err more on the side of semantics but yes this one is important. I do think it's important to explicitely deref because even if it seems obvious to us, it might not be to everyone. You are someone that has an excellent grasp of the language and I do understand the frustration of having to add that extra hoop. That being said, that extra hoop will be very useful for those that only partially understand that feature. We want people to have confidence in running equality on records because they will know of thay have put references in there or not.

**Comment by rricard** [comment]
>  but i don't think we should use pain as a reminder.

I do admit that I prefer solutions that fail fast rather than experiencing weird and hard to explain behavior down the road in a program. I don't think it is pain though, but I can hear that is a subjective thing.

**Comment by littledan** [comment]
I think there's another ergonomic pattern, which is creating a single RefCollection at the top-level of your module (or Realm-wide, even), and using that in all of the functions which manipulate data structures. In my mind, enabling this coding pattern is *the* reason that this proposal has "weak" semantics; none of the examples in the explainer get at this, so it's not surprising if readers are confused.

**Comment by devsnek** [comment]
@littledan imagine A and B come from different modules then.

**Comment by littledan** [comment]
@devsnek Would this be solved by the combination of:
- A built-in per-realm RefCollection to use "by default", so you don't have to worry about which one is which
- Symbols generated by the `ref()` method are specially marked (with a pseudo-internal slot parallel to the description, that you can query) so that you can tell that it makes sense to be dereferenced

**Comment by Jack-Works** [comment]
So why can't we allow Symbol as the key of WeakMap? I don't think the `Symbol.for(...)` case should block that. 🙈

**Comment by littledan** [comment]
@Jack-Works I agree, though I thought this proposal would never get consensus, and this is a lot of what motivated this proposal being the new capability rather than just using WeakMaps. Let's follow up in https://github.com/tc39/ecma262/issues/1194 on that question.

**Comment by devsnek** [comment]
@littledan 

> A built-in per-realm RefCollection

then someone working with message passing and iframes runs into this problem.

> Symbols generated by the ref() method are specially marked

a value being used as a key in some map somewhere is a property of the map, not the value, so I don't think marking symbols makes sense.

**Comment by littledan** [comment]
> then someone working with message passing and iframes runs into this problem.

Yes, well, supporting that sort of use case was why I though it would be good if `ref` supported that second argument of the explicit symbol to provide, for thesesorts of use cases.

> a value being used as a key in some map somewhere is a property of the map, not the value, so I don't think marking symbols makes sense.

RefCollections aren't general-purpose mappings. They generate symbols for specific uses here. So it's not clear to me why these sorts of generalizing invariants would need to hold. (Obviously I'm not suggesting mutating existing Symbols, but rather something that would be set when a Symbol is created.)

**Comment by devsnek** [comment]
> ref supported that second argument of the explicit symbol to provide, for thesesorts of use cases.

But then you're back to the ambiguity case about which map is being used in which case. You also have to deal with the fact that both sides might be using libraries which consume or produce records. If you still need to pass the ref collection between every library boundary because of this use case then the per-realm collection hasn't helped at all.

> RefCollections aren't general-purpose mappings. They generate symbols for specific uses here.

Yes, but symbols are a general purpose value. You can take one of these ref symbols and compose it into other code as a normal symbol, so it would be weird to have a happensToBeUsedAsARefSomewhere flag.

**Comment by littledan** [comment]
> But then you're back to the ambiguity case about which map is being used in which case.

Not necessarily: We could design it to allow you could supply this second argument to the built-in per-Realm RefCollection. You'd only need to supply the argument in the particular cases where you plan to share a new object between Realms, not every single case. And, if the object is unknown, it will be clearly shown as such by the API (returning undefined from deref), not silently failing.

>  so it would be weird to have a happensToBeUsedAsARefSomewhere flag.

It's not about whether it is being used as a ref, it's whether it was built for that. We already sort of have such a queryable flag for whether something was the result of `Symbol.for`; one option would be to just use that one. (Yes, I know there are well-known Symbols, but there are also well-known Objects that will stay alive as long as their Realm...)

These are all tradeoffs. I think it's good to analyze all of them so we could make a balanced choice, but I don't think we've hit a fatal flaw yet.

**Comment by devsnek** [comment]
> You'd only need to supply the argument in the particular cases where you plan to share a new object between Realms

So then every layer has to be aware the value will be moving to another realm (to know to supply this argument). When you consume or produce a value today you don't really have to worry about which realm it was from (instanceof checks aside), 

> It's not about whether it is being used as a ref, it's whether it was built for that. We already sort of have such a queryable flag for whether something was the result of `Symbol.for`

Can you expand? The only way I know of to check is `Symbol.for(s.description) === s`, which bases its truth on a lookup from the map, not any distinct property of `s`. The value you get from `Symbol.for` or `Symbol` bears no inherent information about its "intended" usage, it is just a value, composable into whatever situation you want.

**Comment by littledan** [comment]
Whatever we do for sharing references to objects from primitives will have to indirect through Realms or some other kind of context object, otherwise it would break membranes. Nevertheless, I believe this is useful, and it would be worth it to work out APIs for this case. I agree with you that there is complexity involved; I don't think we're at the point where we can conclude it's fatal yet.

To see whether a symbol came from `Symbol.for`, use the method `Symbol.keyFor` and check whether the return value is undefined.

**Comment by devsnek** [comment]
> Whatever we do for sharing references to objects from primitives will have to indirect through Realms or some other kind of context object, otherwise it would break membranes. Nevertheless, I believe this is useful, and it would be worth it to work out APIs for this case. I agree with you that there is complexity involved; I don't think we're at the point where we can conclude it's fatal yet.

And if we do indirect through the realm, you become unable to use code that doesn't account for that. I don't think we should have to choose between membranes and usable code.

> To see whether a symbol came from Symbol.for, use the method Symbol.keyFor and check whether the return value is undefined.

Symbol.keyFor again distinguishes the symbol using information that is not stored in the symbol. It isn't possible to tell if a symbol is registered without asking the registry.

**Comment by littledan** [comment]
I'm going to sign off for the day after this comment. I think the issues you're raising are good to think through, but I continue to think that we can work through them and find solutions.

> And if we do indirect through the realm, you become unable to use code that doesn't account for that.

Well, what I've been trying to get at is, you'd have an explicit RefCollection constructor, and there'd be a built-in one per Realm. So you'd get to choose. There's definitely some bookkeeping involved, and it'd be good to iterate on API designs based on experience prototyping the proposal, especially to improve ergonomics. Maybe we'll find this is too hard to use, but I don't think we're there yet.

> Symbol.keyFor again distinguishes the symbol using information that is not stored in the symbol. It isn't possible to tell if a symbol is registered without asking the registry.

We can phrase it exactly the same way from an editorial perspective with these kinds of symbols as well. There could be a cross-realm "registry" which is just a set of Symbols which were created along this path, and all you can do is query membership in it. (Or, we could just allow all Symbols which don't come from `Symbol.for`, and reference the same registry. Or, we could make a new primitive type, and not need to query any registry. There are many solutions to this problem; there's no fundamental limitation here, only tradeoffs.)
